swap/swap.go

```diff
@@ -1,98 +1,577 @@
+			// and we actually need to log this error as we are in an async routine; nobody is handling this error for now
+			log.Error("error submitting cheque", "err", err)
+			return
+		}
+		log.Debug("submit tx mined", "receipt", receipt)
+
+		receipt, err = otherSwap.CashChequeBeneficiary(opts, s.backend, s.owner.Contract, big.NewInt(int64(actualAmount)))
+		if err != nil {
+			// TODO: do something with the error
+			// and we actually need to log this error as we are in an async routine; nobody is handling this error for now
+			log.Error("error cashing cheque", "err", err)
+			return
+		}
+		log.Info("Cheque successfully submitted and cashed")
+	}()
+	return err
+}
+
+// processAndVerifyCheque verifies the cheque and compares it with the last received cheque
+// if the cheque is valid it will also be saved as the new last cheque
+func (s *Swap) processAndVerifyCheque(cheque *Cheque, p *Peer) (uint64, error) {
+	if err := cheque.verifyChequeProperties(p, s.owner.address); err != nil {
+		return 0, err
+	}
+
+	lastCheque := s.loadLastReceivedCheque(p)
+
+	// TODO: there should probably be a lock here?
+	expectedAmount, err := s.oracle.GetPrice(cheque.Honey)
+	if err != nil {
+		return 0, err
+	}
+
+	actualAmount, err := cheque.verifyChequeAgainstLast(lastCheque, expectedAmount)
+	if err != nil {
+		return 0, err
+	}
+
+	if err := s.saveLastReceivedCheque(p, cheque); err != nil {
+		log.Error("error while saving last received cheque", "peer", p.ID().String(), "err", err.Error())
+		// TODO: what do we do here? Related issue: https://github.com/ethersphere/swarm/issues/1515
+	}
+
+	return actualAmount, nil
+}
+
+// To be called with mutex already held
+// Caller must be careful that the same balance isn't concurrently read and written by multiple routines
+func (s *Swap) updateBalance(peer enode.ID, amount int64) (int64, error) {
 	//adjust the balance
 	//if amount is negative, it will decrease, otherwise increase
-	s.balances[peer.ID()] += amount
+	balance, exists := s.getBalance(peer)
+	if !exists {
+		return 0, fmt.Errorf("peer %v does not exist", peer)
+	}
+	newBalance := balance + amount
+	s.setBalance(peer, newBalance)
 	//save the new balance to the state store
-	peerBalance := s.balances[peer.ID()]
-	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	err := s.store.Put(balanceKey(peer), &newBalance)
+	if err != nil {
+		return 0, err
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(newBalance, 10))
+	return newBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
-	return err
+// loadBalance loads balances from the state store (persisted)
+// To be called with mutex already held
+// Caller must be careful that the same balance isn't concurrently read and written by multiple routines
+func (s *Swap) loadBalance(peer enode.ID) (err error) {
+	var peerBalance int64
+	if _, ok := s.getBalance(peer); !ok {
```

Comment:
Calling getBalance and setBalance under the accountingLock is protecting the data integrity of the balance. But that also means that balancesLock is not needed.